### ISSUE TITLE
Bug in Twisted Names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11105]
|<img alt="wdy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| wdy reported|
|-|-|
|Trac ID|trac#1581|
|Type|defect|
|Created|2006-03-16 08:14:43Z|

There seems to be a bug with Twisted Names.

ver 2.2.0

In twisted/names/dns.py,
dns.Query
```
    def __init__(self, name='', '''type=A, cls=IN'''):  
        """
        @type name: C{str}
        @param name: The name about which to request information.

        @type type: C{int}
        @param type: The query type.

        @type cls: C{int}
        @param cls: The query class.
        """
        self.name = Name(name)
        self.type = type
        self.cls = cls
```

the parameter: first type, then cls

In twisted/names/root.py,
```
    def discoveredAuthority(self, auth, name, cls, type, timeout):
        from twisted.names import client

        q = dns.Query(name, '''cls, type''')
        #here  first cls, then type

        r = client.Resolver(servers=[(auth, dns.PORT)])
        d = r.queryUDP([q], timeout)
        d.addCallback(r.filterAnswers)
        return d
```
In twisted/names/cache.py,
cache.CacheResolver,
```
    def _lookup(self, name, cls, type, timeout):
        now = time.time()
        q = dns.Query(name, '''cls, type''')
        #here  first cls, then type
        try:
            when, (ans, auth, add) = self.cache[q]
        except KeyError:
            if self.verbose > 1:
                log.msg('Cache miss for ' + repr(name))
            return defer.fail(failure.Failure(dns.DomainError(name)))
        else:
            ............
```

<details><summary>Searchable metadata</summary>

```
trac-id__1581 1581
type__defect defect
reporter__wdy wdy
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1142496883000000 1142496883000000
changetime__1142566069000000 1142566069000000
version__2_0 2.0
owner__ 
cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Please don't put random things into the keywords field.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to Jerub|
|-|-|

Fix in cls-type-1581, pending review.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @exarkun|
|-|-|

Please commit this now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16331]) Merge cls-type-1581

Author: exarkun
Reviewer: jerub
Fixes [#11105](https://github.com/twisted/twisted/issues/11105)

This adds a test for basic cache lookup behavior and fixes a bug in that
code as well.  It also fixes the same bug in the root resolver, but there
are no tests for that code (which is probably okay, because I doubt it
works very well anyway).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
